### ISSUE TITLE
Non-OCLC numbers generating HathiTrust links

### ISSUE THREAD

**Issue by ghost** [UMNLibraries/primo-explore-hathitrust-availability#2]
We had an issue where the updateHathiTrustAvailability function generated a link from a non-OCLC 035 field.

We solved the issue by fixing the record (it contained data that none of us recognized), but you might want to filter the updateHathiTrustAvailability function to only search on pnx.addata.oclcid with a (ocolc) prefix.

**Comment by gpeterso** [comment]
Hi @michaelsse, 

Thanks for the suggestion. 

Are you saying that the `addata.oclcid` values in your PNX records contain the original 035 prefixes (such as "OCoLC")? This surprises me a bit because Primo's normalization rules template for MARC records removes this prefix explicitly (*Take string (use reg. exp.)* `\(OCoLC\)[^\d]*(\d+)`), which results in numeric-only `addata.oclcid` values: 

```javascript
addata: {
 ...
  oclcid: [
  "1586310",
  "7417753",
  "47076528"
  ], 
...
```

At least, that's the way "classic" Primo works.... Are you on VE, by chance?

Jeff

**Comment by gibsonjc** [comment]
Hello @gpeterso ! Sorry to comment on a closed issue; I can open a new one if you prefer!
We are relatively new to Alma/Primo VE (never used Primo Classic) and love your add-on, but we have noticed this problem is still occurring with unprefixed 035 fields.  For example, a problem record we found yesterday has several 035 fields like this:
035	##$a (OCoLC)1164818135 
035	##$a 16696409 
035	##$z (OCoLC)1202530655 
035	##$a (OCoLC)on1164818135 

The add-on is matching on the one with no prefix (16696409, which is some kind of vendor number) and creating a link to a completely different title in HathiTrust that actually has OCLC number 16696409.  We can fix this by editing the individual bib record directly to remove that 035 and try to avoid unprefixed 035s in future records to the best of our ability, but I suspect there are many more unprefixed 035s out there in other bib records and they can't be prevented 100% of the time.  It would be great if the add-on could limit its API search to only 035s with OCLC numbers as indicated by the presence of prefixes, if that makes sense.  

If you'd like to see this problem in action in our Primo VE, I'd be happy to send you a link to it and hold off on fixing it for a time.  Thank you for your work on this outstanding add-on!
-Jessica

**Comment by gpeterso** [comment]
Hi @gibsonjc , 

Thanks for reaching out. Yeah, if you could provide a link to a sample record, that would be super helpful. 

Jeff

**Comment by gibsonjc** [comment]
Thanks so much for looking into it, @gpeterso !
Here's a link in EIU's Primo VE for "Yemen : what everyone needs to know" that we haven't deleted the unprefixed 035 from yet:

https://i-share-eiu.primo.exlibrisgroup.com/permalink/01CARLI_EIU/1qndmrq/alma9917569013005833

Library staff were immediately suspicious because it's a book from 2021 with a HathiTrust link. Upon further examination, we discovered the bib 035 with no prefix that matches the OCLC number for "Oregon plans and proceedings in the adult and nursery school educational program, 1935-36" which is in HathiTrust.

Let me know if you need any other information!
thanks again,
-Jessica
Consortium of Academic and Research Libraries in Illinois (CARLI)

**Comment by gpeterso** [comment]
Thanks for the example, @gibsonjc !

It looks like there might be an issue with the way VE is mapping OCLC numbers to the [PNX record](https://i-share-eiu.primo.exlibrisgroup.com/discovery/fulldisplay?context=L&vid=01CARLI_EIU:CARLI_EIU&search_scope=MyInstitution&tab=LibraryCatalog&docid=alma9917569013005833&showPnx=true). Specifically, the `addata.oclcid` field consists of the following values: ` [ "(ocolc)1164818135", "16696409", "(ocolc)1202530655", "(ocolc)on1164818135" ]`. 

I think I could come up with a workaround for VE, but ideally Ex Libris would fix this because mapping an unprefixed `035 $a` value (in this case `035 ##$a 16696409`) to a field named "oclcid" is incorrect and likely to cause problems elsewhere. 

Would you be willing to open a SF case with Ex Libris? If they don't seem interested in fixing it quickly, I'll go ahead and implement a workaround. 

-- 
Jeff

**Comment by gibsonjc** [comment]
Thanks @gpeterso , I can certainly create a new SF Case and give it a try!  Would you mind if I CCed you on the new Case so you can see and participate in any discussion directly?  I'm a System Admin and a Cataloger, but not a coder, and I'm not terribly familiar with the PNX record!
-Jessica

**Comment by gpeterso** [comment]
Of course. Feel free to cc me on the case @gibsonjc

**Comment by gibsonjc** [comment]
Thank you @gpeterso ! Case #00982125. Fingers crossed!

**Comment by dlingley** [comment]
Hi @gibsonjc, did you ever hear back from ExL on the addata.oclcid field problem?  Or come up with any workaround?

**Comment by gibsonjc** [comment]
Hi @dlingley - Our Case is now 05312527 - Unprefixed 035 data in Primo VE's oclcid PNX field - Tier 1 replicated it and sent it to Tier 2, who replied "Thank you for reporting this issue with Primo VE's oclcid values. This issue has been reported by a few other customers as well. I have linked your Case to a Development Issue and shared your scenario with the developers." The Case is currently "Pending – Development" status.  Unfortunately no, we don't have a workaround and some records are still linking out to the incorrect HathiTrust item.

Also tagging creator @gpeterso since ExL doesn't seem interested in fixing it very quickly, and he mentioned earlier in the thread that he might have an idea if they didn't.  ;-)

**Comment by alicetippit** [comment]
Hi @dlingley chiming in with our workaround here at Northwestern. We set the plugin to use a display field that only has the OCLC number in it. This is a field that we added so we could display the OCLC number in the full record view.

Here is a link to one of our records, with the OCLC field in the Details section: https://search.library.northwestern.edu/permalink/01NWU_INST/h04e76/alma9918180424202441

**Comment by dlingley** [comment]
Thanks for sharing all.  I updated and escalated my ticket too... 

Also tweaked the regex code on line 140: return value.match(/^(\(ocolc\))**?**\d+$/i); 

Just removing the ? makes the (ocolc) string be required: return value.match(/^(\(ocolc\))\d+$/i);

Gonna leave it that way for now until this gets fixed by ExL.  This change broke the hathi entries from CDI because none of them have the (ocolc) prefix.  The online link for Hathi CDI records are still there however, so going to live with this for now.  Not having the incorrect link show up for local records is worth this inconsistency in user experience for now.

Seems ridiculous ExL can't modify the addata oclc id rule to be similar to the rules for the search and identifier display sections, which both require 035a: https://knowledge.exlibrisgroup.com/Primo/Product_Documentation/020Primo_VE/Primo_VE_(English)/120Other_Configurations/Mapping_to_the_Display%2C_Facets%2C_and_Search_Sections_in_the_Primo_VE_Record#:~:text=024%2D2%20a-,OCLC,035%20a%20(Only%20if%20the%20string%20OCoLC%20exists%20in%20the%20value.),-PUBNUM

https://knowledge.exlibrisgroup.com/Primo/Product_Documentation/020Primo_VE/Primo_VE_(English)/120Other_Configurations/Mapping_to_the_Display%2C_Facets%2C_and_Search_Sections_in_the_Primo_VE_Record#:~:text=e%2C%20y%2C%20z-,OCLC,035%20a%20(Only%20if%20the%20string%20OCoLC%20exists%20in%20the%20value.),-Language

Also wondering why we don't have access to edit the addata norm rules.  I added that question to my case as well.

**Comment by gpeterso** [comment]
Hi all, 

Thanks for reminding me of this problem. I just published version 2.7.1, which includes a VE-specific fix for OCLC IDs. This will need to be updated when Ex Libris decides to fix the metadata. If that does occur and you need a quick fix, you could just revert to 2.7.0. 

Jeff
